cpp/src/arrow/io/buffered_test.cc

```diff
@@ -491,6 +491,30 @@ TEST_F(TestBufferedInputStream, BufferSizeLimit) {
   }
 }
 
+TEST_F(TestBufferedInputStream, PeekPastBufferedBytes) {
+  // GH-43949: Peek and SetBufferSize should not affect the
+  // buffered bytes.
+  MakeExample1(/*buffer_size=*/10, default_memory_pool(), /*raw_read_bound=*/15);
+  ASSERT_OK_AND_ASSIGN(auto bytes, buffered_->Read(9));
+  EXPECT_EQ(std::string_view(bytes->data_as<char>(), bytes->size()),
+            kExample1.substr(0, 9));
+  ASSERT_EQ(1, buffered_->bytes_buffered());
+  ASSERT_EQ(10, buffered_->buffer_size());
+  ASSERT_OK_AND_ASSIGN(auto view, buffered_->Peek(3));
+  EXPECT_EQ(view, kExample1.substr(9, 3));
+  ASSERT_EQ(3, buffered_->bytes_buffered());
+  ASSERT_EQ(12, buffered_->buffer_size());
+  ASSERT_OK_AND_ASSIGN(view, buffered_->Peek(10));
+  EXPECT_EQ(view, kExample1.substr(9, 6));
```

Comment:
Why is it not `kExample1.substr(9, 10)`?

Reply:
Ah, I understand: it's because of the read bound.

Reply:
Perhaps add a comment
```suggestion
  // Peek() cannot go past the `raw_read_bound`
  EXPECT_EQ(view, kExample1.substr(9, 6));
```